salt/_modules/metalk8s_solutions.py

```diff
@@ -202,52 +201,110 @@ def list_solution_images(mountpoint):
     return {
-        'kind': SOLUTION_CONFIG_KIND,
-        'apiVersion': SOLUTION_CONFIG_APIVERSIONS[0],
-        'operator': {
-            'image': {
-                'name': '{}-operator'.format(name),
-                'tag': version,
+        'spec': {
+            'operator': {
+                'image': {
+                    'name': '{}-operator'.format(name),
+                    'tag': version,
+                },
             },
-        },
-        'ui': {
-            'image': {
-                'name': '{}-ui'.format(name),
-                'tag': version,
+            'ui': {
+                'image': {
+                    'name': '{}-ui'.format(name),
+                    'tag': version,
+                },
             },
+            'images': list_solution_images(mountpoint),
+            'customApiGroups': [],
         },
-        'images': list_solution_images(mountpoint),
-        'customApiGroups': [],
     }
 
 
-def read_solution_config(mountpoint, name, version):
-    log.debug('Reading Solution config from %r', mountpoint)
-    config = _default_solution_config(mountpoint, name, version)
-    config_path = os.path.join(mountpoint, 'config.yaml')
-
-    if not os.path.isfile(config_path):
-        log.debug('Solution mounted at %r has no "config.yaml"', mountpoint)
-        return config
+def read_solution_manifest(mountpoint):
+    log.debug('Reading Solution manifest from %r', mountpoint)
+    manifest_path = os.path.join(mountpoint, SOLUTION_MANIFEST)
 
-    with salt.utils.files.fopen(config_path, 'r') as stream:
-        provided_config = salt.utils.yaml.safe_load(stream)
+    if not os.path.isfile(manifest_path):
+        raise CommandExecutionError(
+            'Solution mounted at "{}" has no "{}"'
+            .format(mountpoint, SOLUTION_MANIFEST)
+        )
 
-    provided_kind = provided_config.pop('kind', None)
-    provided_api_version = provided_config.pop('apiVersion', None)
+    with salt.utils.files.fopen(manifest_path, 'r') as stream:
+        manifest = salt.utils.yaml.safe_load(stream)
 
     if (
-        provided_kind != SOLUTION_CONFIG_KIND or
-        provided_api_version not in SOLUTION_CONFIG_APIVERSIONS
+        manifest.get('kind') != SOLUTION_MANIFEST_KIND or
+        manifest.get('apiVersion') not in SOLUTION_MANIFEST_APIVERSIONS
     ):
         raise CommandExecutionError(
-            'Wrong apiVersion/kind for {}'.format(config_path)
+            'Wrong apiVersion/kind for {}'.format(manifest_path)
         )
 
-    salt.utils.dictupdate.update(config, provided_config)
-    return config
+    info = _archive_info_from_manifest(manifest)
+    default_manifest = _default_solution_manifest(
+        mountpoint, info['name'], info['version']
+    )
+
+    manifest = salt.utils.dictupdate.merge(
+        default_manifest, manifest, strategy='recurse'
+    )
+
+    return manifest, info
+
+
+def _archive_info_from_manifest(manifest):
+    info = {}
```

Comment:
As a general comment, this way of writing methods that build and return an object, although widespread in Salt modules, makes the code quite hard to read. You need to parse through each branch to understand what the contents of an object will be when it's returned. Sprinkle some early returns in there and it's quickly becoming a mess.

Instead, I'd suggest we only build the structure in the final `return` statement (even in an early `return`?). Something like this:

```python
def _archive_info_from_manifest(manifest):
    name = manifest.get('metadata', {}).get('name')
    version = manifest.get('spec', {}).get('version')

    if any(key is None for key in [name, version]):
        raise CommandExecutionError(
            'Missing mandatory key(s) in Solution "{}": must provide '
            '"metadata.name" and "spec.version"'
            .format(SOLUTION_MANIFEST)
        )

    display_name = manifest.get('annotations', {}).get(
        'solutions.metalk8s.scality.com/display-name', name
    )

    return {
        'name': name,
        'version': version,
        'display_name': display_name,
        'id': '-'.join([name, version]),
    }
```

Reply:
As you prefer, I did this because as you said, most of the time it's the approach used in Salt modules.

Reply:
Consistently bad is still bad :P I'd rather have us change this "standard" progressively by checking for this pattern in reviews than do a full refactor on thousands of lines in a few months...